launch/launch/launch_service.py

```diff
@@ -338,28 +339,39 @@ def _on_exception(loop, context):
                         ret = await self._shutdown(reason='idle', due_to_sigint=False)
                         assert ret is None, ret
                         continue
-                    process_one_event_task = this_loop.create_task(self._process_one_event())
-                    if self.__shutting_down:
-                        # If shutting down and idle then we're done.
-                        if is_idle:
+
+                    # Stop running if we're shutting down and there's no more work
+                    if self.__shutting_down and is_idle:
+                        if (
+                            process_one_event_task is not None and
+                            not process_one_event_task.done()
+                        ):
                             process_one_event_task.cancel()
+                        break
```

Comment:
are the shutdown event handlers going to be executed if we just break here?
should something be awaited after shutdown happens?

I have to think a bit more ... I haven't figured out what the previous code was doing.

Reply:
I think this logic is equivalent to the code before:

https://github.com/ros2/launch/blob/f3f0d69d8b43eecbcf968ee06611960a33be6313/launch/launch/launch_service.py#L342-L346

If there are still events in the queue, then `is_idle` will be `False` and we'll continue to handle them.

Reply:
you're right